File: apps/web/src/server/jobs/booking-created.ts
/**
 * `booking.created` — tell one recipient that a booking exists.
 * One job per recipient (see `contracts/jobs.ts`), so this handler always
 * sends exactly one message and a retry re-sends only to the party that failed.
 *
 * Reached as a QStash delivery to `POST /api/jobs/booking.created`; the
 * dispatch layer (`run.ts`) has already validated the payload.
 */

import { type BookingCreatedJob, isDemoOrganizerId } from '@repo/contracts'

import { issueLoginLink } from '../auth/login-link'
import { getPostHog } from '../posthog'
import type { JobsEnv } from './env'
import { cabinetSlotPath, loginLinkUrl, manageBookingUrl } from './links'
import { getNotificationContext } from './notification-context'
import { sendMessage } from './telegram/client'
import {
  bookingCreatedForGuest,
  bookingCreatedForOrganizer,
  notificationLocale,
} from './telegram/templates'

import 'server-only'

export async function handleBookingCreated(
  env: JobsEnv,
  data: BookingCreatedJob,
): Promise<void> {
  const { bookingId, recipient } = data

  const context = await getNotificationContext(bookingId)
  if (!context) {
    console.warn(`[booking.created] booking ${bookingId} no longer exists — skipping`)
    return
  }

  if (isDemoOrganizerId(context.organizer.id)) {
    console.warn(`[booking.created] refusing to notify the demo organizer (${bookingId})`)
    return
  }

  const ph = getPostHog()

  if (recipient === 'organizer') {
    const token = await issueLoginLink(context.organizer.id, cabinetSlotPath(context.slot.id))

    const message = bookingCreatedForOrganizer(
      context,
      loginLinkUrl(env.appUrl, token),
      notificationLocale(recipient, context),
    )

    await sendMessage(env.telegramBotToken, {
      chatId: context.organizer.messengerId,
      text: message.text,
      button: message.button,
    })
    ph?.capture({
      distinctId: context.organizer.id,
      event: 'notification_sent',
      properties: { queue: 'booking.created', recipient, bookingId },
    })
    return
  }

  const message = bookingCreatedForGuest(
    context,
    manageBookingUrl(env.appUrl, context.booking.manageToken),
    notificationLocale(recipient, context),
  )

  await sendMessage(env.telegramBotToken, {
    chatId: context.booking.guestMessengerId,
    text: message.text,
    button: message.button,
  })
  ph?.capture({
    distinctId: context.organizer.id,
    event: 'notification_sent',
    properties: { queue: 'booking.created', recipient, bookingId },
  })
}

File: apps/web/pkg/jobs/booking_created.go
package jobs

import (
	"context"

	"countmein/pkg/auth"
	"countmein/pkg/contracts"
	"countmein/pkg/db"
	"countmein/pkg/logx"
)

// booking.created — tell one recipient that a booking exists. One job
// per recipient (see contracts/jobs.go), so this handler always sends
// exactly one message and a retry re-sends only to the party that
// failed. Reached as a QStash delivery to POST /api/jobs/booking.created;
// the dispatch layer (run.go) has already validated the payload.

// HandleBookingCreated notifies one recipient about a fresh booking.
// PostHog captures from the TS handler are not ported (no SDK in the
// dependency set); delivery logging covers the remainder.
func HandleBookingCreated(ctx context.Context, env Env, job contracts.BookingCreatedJob) error {
	booking, slot, service, organizer, err := db.GetBookingChain(ctx, job.BookingID)
	if err != nil {
		return err
	}
	// Deliberately a fresh read at send time: a job that waited out a
	// retry backoff must render the booking as it is now, not as it was
	// when the transaction committed.
	if booking == nil {
		logx.Info("booking no longer exists — skipping", map[string]any{
			"queue": contracts.QueueBookingCreated, "bookingId": job.BookingID,
		})
		return nil
	}
	view := BookingView{Booking: *booking, Slot: *slot, Service: *service, Organizer: *organizer}

	// Demo bookings never reach a chat (ADR-010).
	if contracts.IsDemoOrganizerID(organizer.ID) {
		logx.Info("refusing to notify the demo organizer", map[string]any{
			"queue": contracts.QueueBookingCreated, "bookingId": job.BookingID,
		})
		return nil
	}

	if job.Recipient == contracts.RecipientOrganizer {
		// Minted per send attempt: a retry mints a fresh token and the
		// abandoned one simply expires, so a delivered message never
		// carries a button already spent by an earlier attempt.
		token, err := auth.IssueLoginLink(ctx, organizer.ID, CabinetSlotPath(slot.ID))
		if err != nil {
			return err
		}
		locale := NotificationLocale(contracts.RecipientOrganizer, view)
		message := BookingCreatedForOrganizer(view, LoginLinkURL(env.AppURL, token), locale)
		return SendMessage(ctx, env.TelegramBotToken, organizer.MessengerID, message.Text, message.Button)
	}

	locale := NotificationLocale(contracts.RecipientGuest, view)
	message := BookingCreatedForGuest(view, ManageBookingURL(env.AppURL, booking.ManageToken), locale)
	return SendMessage(ctx, env.TelegramBotToken, booking.GuestMessengerID, message.Text, message.Button)
}
